cmd/terraform-mcp-server/instructions.md

```diff
@@ -0,0 +1,76 @@
+
+## Always Available Tools
+
+### Registry Tools (Always Available)
+- **Provider Discovery**: get_latest_provider_version (if unavailable in code) → search_providers → get_provider_details
+- **Module Discovery**: get_latest_module_version (if unavailable in code) → search_modules → get_module_details
+- **Policy Discovery**: search_policies → get_policy_details
+- Use these to ensure generated code uses current versions and follows best practices
+
+## HCP Terraform/TFE Tools (Requires TFE_TOKEN)
+
+### Private Registry Tools
+- search_private_providers → get_private_provider_details
+- search_private_modules → get_private_module_details
+- Priority: Check private registries first when token present, public as fallback
+
+### Workspace Management
+- **Discovery**: search_workspaces (empty query returns all) → get_workspace_details
+- **Operations**: create_workspace, update_workspace, delete_workspace_safely
+- delete_workspace_safely only works if workspace has no managed resources
+
+### Run Execution
+- **Discovery**: search_run (empty query returns all) → get_run_details (supports json output)
+- **Operations**: create_run → apply_run OR discard_run OR cancel_run
+- **Monitoring**: get_plan_details/get_plan_logs for plans, get_apply_details/get_apply_logs for applies
+- Always check run status before attempting operations
+
+### Variable Management
+**Workspace Variables**:
+- search_workspace_variables (empty query returns all)
+- create_workspace_variable, update_workspace_variable, delete_workspace_variable
+
+**Variable Sets** (for sharing across workspaces/projects):
+- search_variable_sets → get_variable_set_details
+- create_variable_set, update_variable_set, delete_variable_set
+- create_variable_in_variable_set, update_variable_in_variable_set, delete_variable_from_variable_set
+- attach/detach_variable_set_to_workspaces, attach/detach_variable_set_to_projects
+
+## Workflow Patterns
+
+**Code Generation**:
+1. search_modules/search_providers for available resources
+2. get_latest_provider_version is no version available in existing code
+3. get_module_details for module requirements
+4. Generate code with discovered constraints
+
+**Run Management**:
+1. search_workspaces → select target
+2. create_run → get_run_details to monitor
+3. get_plan_details/logs to review changes
+4. User confirmation → apply_run OR discard_run
+
+**Variable Configuration**:
+1. search_workspace_variables to check existing
+2. create/update_workspace_variable as needed
+3. For multi-workspace: create_variable_set → attach_variable_set_to_workspaces
+
+## Error Handling
+- Registry failures: Try private first (if token), fallback to public
+- Run failures: Check get_run_details, get_plan_details and logs before retry
+- Variable conflicts: search_workspace_variables first to avoid duplicates
+
+## Security Notes
+- Never expose TFE_TOKEN in outputs
```

Comment:
The agent will never have direct access to it though, right?

Reply:
this is resolved in the latest commit